File: app/Models/Order.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use App\Models\PaymentReceipt;

class Order extends Model
{
    use HasFactory;

    protected $fillable = [
        'buyer_id',
        'order_number',
        'total_amount',
        'status',
        'delivery_address',
        'phone',
        'payment_method',
        'notes',
    ];

    protected $casts = [
        'total_amount' => 'decimal:2',
    ];

    /**
     * Get the buyer that owns the order
     */
    public function buyer()
    {
        return $this->belongsTo(User::class, 'buyer_id');
    }

    /**
     * Get the order items for the order
     */
    public function orderItems()
    {
        return $this->hasMany(OrderItem::class);
    }

    /**
     * Get the crops associated with this order
     */
    public function crops()
    {
        return $this->belongsToMany(Crop::class, 'order_items')
                    ->withPivot(['quantity', 'price_per_kg', 'total_price']);
    }

    /**
     * Scope a query to only include pending orders
     */
    public function scopePending($query)
    {
        return $query->where('status', 'pending');
    }

    /**
     * Scope a query to only include confirmed orders
     */
    public function scopeConfirmed($query)
    {
        return $query->where('status', 'confirmed');
    }

    /**
     * Scope a query to only include delivered orders
     */
    public function scopeDelivered($query)
    {
        return $query->where('status', 'delivered');
    }

    /**
     * Generate a unique order number
     */
    public static function generateOrderNumber()
    {
        $latestOrder = self::latest()->first();
        $orderNumber = 'ORD' . date('Ymd') . '001';
        
        if ($latestOrder) {
            $lastOrderNumber = $latestOrder->order_number;
            $lastSequence = intval(substr($lastOrderNumber, -3));
            $orderNumber = 'ORD' . date('Ymd') . str_pad($lastSequence + 1, 3, '0', STR_PAD_LEFT);
        }
        
        return $orderNumber;
    }

    /**
     * Get the formatted total amount
     */
    public function getFormattedTotalAttribute()
    {
        return 'TZS' . number_format($this->total_amount, 2);
    }

    /**
     * Get payment receipts for this order
     */
    public function paymentReceipts()
    {
        return $this->hasMany(PaymentReceipt::class, 'order_id');
    }

    /**
     * Get the status badge HTML
     */
    public function getStatusBadgeAttribute()
    {
        $badges = [
            'pending' => '<span class="badge bg-warning">Pending</span>',
            'pending_payment' => '<span class="badge bg-secondary">Pending Payment</span>',
            'payment_rejected' => '<span class="badge bg-danger">Payment Rejected</span>',
            'confirmed' => '<span class="badge bg-info">Confirmed</span>',
            'delivered' => '<span class="badge bg-success">Delivered</span>',
            'cancelled' => '<span class="badge bg-danger">Cancelled</span>',
        ];
        
        return $badges[$this->status] ?? $badges['pending'];
    }
}

File: farmmarket_mobile/lib/models/order.dart
import 'crop.dart';

class Order {
  final int id;
  final String orderNumber;
  final int buyerId;
  final int farmerId;
  final double totalAmount;
  final String status;
  final String deliveryAddress;
  final String paymentMethod;
  final String? paymentReceipt;
  final String? trackingNumber;
  final String? notes;
  final String createdAt;
  final String updatedAt;
  final Buyer buyer;
  final Farmer farmer;
  final List<OrderItem> items;

  Order({
    required this.id,
    required this.orderNumber,
    required this.buyerId,
    required this.farmerId,
    required this.totalAmount,
    required this.status,
    required this.deliveryAddress,
    required this.paymentMethod,
    this.paymentReceipt,
    this.trackingNumber,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
    required this.buyer,
    required this.farmer,
    required this.items,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: json['id'],
      orderNumber: json['order_number'],
      buyerId: json['buyer_id'],
      farmerId: json['farmer_id'],
      totalAmount: double.parse(json['total_amount'].toString()),
      status: json['status'],
      deliveryAddress: json['delivery_address'],
      paymentMethod: json['payment_method'],
      paymentReceipt: json['payment_receipt'],
      trackingNumber: json['tracking_number'],
      notes: json['notes'],
      createdAt: json['created_at'],
      updatedAt: json['updated_at'],
      buyer: Buyer.fromJson(json['buyer']),
      farmer: Farmer.fromJson(json['farmer']),
      items: (json['items'] as List<dynamic>?)
          ?.map((item) => OrderItem.fromJson(item))
          .toList() ?? [],
    );
  }

  String get statusDisplay {
    switch (status) {
      case 'pending':
        return 'Pending';
      case 'confirmed':
        return 'Confirmed';
      case 'shipped':
        return 'Shipped';
      case 'delivered':
        return 'Delivered';
      case 'cancelled':
        return 'Cancelled';
      default:
        return status;
    }
  }
}

class OrderItem {
  final int id;
  final int orderId;
  final int cropId;
  final int farmerId;
  final int quantity;
  final double pricePerUnit;
  final double totalPrice;
  final String createdAt;
  final String updatedAt;
  final Crop? crop;

  OrderItem({
    required this.id,
    required this.orderId,
    required this.cropId,
    required this.farmerId,
    required this.quantity,
    required this.pricePerUnit,
    required this.totalPrice,
    required this.createdAt,
    required this.updatedAt,
    this.crop,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      id: json['id'],
      orderId: json['order_id'],
      cropId: json['crop_id'],
      farmerId: json['farmer_id'],
      quantity: json['quantity'],
      pricePerUnit: double.parse(json['price_per_unit'].toString()),
      totalPrice: double.parse(json['total_price'].toString()),
      createdAt: json['created_at'],
      updatedAt: json['updated_at'],
      crop: json['crop'] != null ? Crop.fromJson(json['crop']) : null,
    );
  }
}

class Buyer {
  final int id;
  final String name;
  final String email;
  final String phone;

  Buyer({
    required this.id,
    required this.name,
    required this.email,
    required this.phone,
  });

  factory Buyer.fromJson(Map<String, dynamic> json) {
    return Buyer(
      id: json['id'],
      name: json['name'],
      email: json['email'],
      phone: json['phone'],
    );
  }
}

class Farmer {
  final int id;
  final String name;
  final String email;
  final String phone;

  Farmer({
    required this.id,
    required this.name,
    required this.email,
    required this.phone,
  });

  factory Farmer.fromJson(Map<String, dynamic> json) {
    return Farmer(
      id: json['id'],
      name: json['name'],
      email: json['email'],
      phone: json['phone'],
    );
  }
}

